### PR TITLE
Refactor Associations::Preloader::Association module

### PR DESCRIPTION
**Pull request by wentaoxu415**
**Disclaimer**
This is a pretty deep rabbit hole and I recommend our team to do the following 2 things in the near future after this PR:
1) Ditch the `primary-composite-key` gem as soon as we can 
2) Improve our data engineering practice on how we standardize the join keys between different tables.

**Context**
My investigation of why preload was failing for Elastic documents creation led me to this specific module within the `primary-composite-key` gem called `Associations::Preloader::Association`. Basically, the code in this preload module had cases where it can fail to look up the record from hashmap because of the differences in what the DB and Rails considers to be the same join key.

It looks like during the Rails 5 upgrade, @imkirnos has updated our DB to use `utf8_general_ci`, and along with that change, the DB started to also become accent insensitive. 

This accent insensitiveness makes up for the lack of our good data engineering practice around how we introduce standardized key. For example, in the `LinkedinUser` table, we have a linkedin_permalink of `in/tuấn-tô-thanh-49132198` (notice the Vietnamese phonetic letters), but in the `LinkedinAllUseRanking` table, we have a linkedin_permalink of `in/tuan-to-thanh-49132198` (all standard English letters). With the DB being accent insensitive, it allows the join to associate these kinds of records with keys made up of same letters, but with different accent letters.

However, back in Rails / the `primary-composite-key` gem, the preloader association still attempts to do conventional string comparison based on string bytes. Therefore, its behavior out of the box is that it would treat keys such as `in/tuấn-tô-thanh-49132198` and `in/tuan-to-thanh-49132198` to be different and eventually result in lookup failure from hashmap and cause `nil` error.

As a workaround, @imkirnos had introduced some transliteration step to convert the accented keys back to standard ASCII keys as much as possible in the recent commits.  Despite this workaround, there was still an edge case where the lookup failure could happen such as the case where the transliteration does not recognize some letters and convert them to `?` letter which causes lookup failure. Moreover, there was an additional source of bug in which the step to link the associated records back to the primary / owner record could silently fail in cases where the primary / owner record has accented key and the associated records has standard ASCII key.

Lastly, as one last callout, a similar error also occurs for cases where one of the join keys had a Zero Width Space character. This case also warranted a bug fix.

**Solution**
The proposed solution was multi-pronged and it consisted of the following changes:
- Improve the robustness of the record lookup by attempting to look up using the following 3 keys in the waterfall manner:
    - original key
    - standardized key
    - transliterated key
 - Introduce the improved record look up for the step to associate records to owners so that the record look up is robust in both directions (going from owner to associated record(s), as well as associated record to owner(s)
 - Remove Zero Width Space character
 - Fix the bug in the `transliterate_keys`
 - Introduce some preset mapping for transliteration to improve transliteration success rate (introduced in [this PR](https://github.com/Tsquare/beacon/pull/3703/files) for the main Beacon)

### CODE REVIEW

**Comment by kevchoi** [comment]
fyi, for what it's worth, our database has always been case insensitive and thus accent insensitive; see comment & the safe navigator operator addition in lib/composite_primary_keys/associations/preloader/association.rb#51 (i think ilya manually applied the code changes when upgrading the composite primary key gems which mangled git history, but it was mostly howard that made the transliteration hacks to the cpk gem and i wrote the comment and added the safe navigator)

definitely agree with you that we should remove cpk as soon as possible, especially all preloads and joins are routed through this package and not rails

i made the decision to intentionally silence matching issues (safe navigator) because i assumed the ocurrence of accents was negligible compared to the code complexity/performance hit of all string comparisons, transliterations, etc that happen for any type of preload.

tldr; it looks like you're going for robustness, which is fine with me. but will this fix the preloader exceptions we're seeing that's blocking elastic export? otherwise, lgtm
